File: chapter 1/python/1.6.py
# 1.6
# implement a method to perform basic string compression using the counts of repeated characters
# e.g. string 'aabcccccaaa' would become a2b1c5a3
# if the compressed string would not become shorter than the original string, return original string
# can assume the string has only uppercase and lowercase letters

def StringCompression(string):
    if len(string) > 1:
        newString = string[0]
        charCount = 1

        for i in range(1, len(string)):
            if string[i] == string[i-1]:
                charCount += 1
            if string[i] != string[i-1]:
                newString += str(charCount)
                charCount = 1
                newString += string[i]
            if(i == len(string)-1):
                newString += str(charCount)

        if len(newString) < len(string):
            return newString
        else:
            return string
    else:
        return "String cannot be compressed"

string1 = "aaabbbbbccddd"
string2 = "abcd"
string3 = "a"
string4 = "aaabccdddee"

print(StringCompression(string4))

File: chapter 1/js/1.6.js
// 1.6 String Compression: Implement a method to perform basic string compression using the counts
// of repeated characters. For example, the string aabcccccaaa would become a2blc5a3, If the
// "compressed" string would not become smaller than the original string, your method should return
// the original string. You can assume the string has only uppercase and lowercase letters (a - z). 
// I: string
// O: string
// C: compressed str cannot be larger than original
// E: empty string

function compressStr(str){
    if (str.length == 0){
        return null;
    }
    let newStr = "";
    let count = 1;
    let char = str[0];
    for (let i =1; i<=str.length; i++){
        if(char != str[i]){
            newStr = newStr + char + count;
            char = str[i];
            count = 1;
        } else {
            count ++;
        }
    }
    if (str.length <= newStr.length){
        return str;
    }
    return newStr;
}

// console.log(compressStr("aabcccccaaa"))
// console.log(compressStr("bbbbbbccaasst"))